app/assets/javascripts/discourse/app/controllers/login-page.js

```diff
@@ -1,5 +1,410 @@
+            window.location.assign(ssoDestinationUrl);
+          } else if (destinationUrl) {
+            removeCookie("destination_url");
+            window.location.assign(destinationUrl);
+          } else {
+            window.location.reload();
+          }
+        } else {
+          this.dialog.alert(authResult.error);
+        }
+      }
+    } catch (e) {
+      popupAjaxError(e);
+    }
+  }
+
+  @action
+  preloadLogin() {
+    const prefillUsername = document.querySelector(
+      "#hidden-login-form input[name=username]"
+    )?.value;
+    if (prefillUsername) {
+      this.loginName = prefillUsername;
+      this.loginPassword = document.querySelector(
+        "#hidden-login-form input[name=password]"
+      ).value;
+    } else if (cookie("email")) {
+      this.loginName = cookie("email");
+    }
+  }
+
+  @action
+  securityKeyCredentialChanged(value) {
+    this.securityKeyCredential = value;
+  }
+
+  @action
+  flashChanged(value) {
+    this.flash = value;
+  }
+
+  @action
+  flashTypeChanged(value) {
+    this.flashType = value;
+  }
+
+  @action
+  loginNameChanged(event) {
+    this.loginName = event.target.value;
+  }
+
+  @action
+  showCreateAccount(createAccountProps = {}) {
+    if (this.site.isReadOnly) {
+      this.dialog.alert(I18n.t("read_only_mode.login_disabled"));
+    } else {
+      this.handleShowCreateAccount(createAccountProps);
+    }
+  }
+
+  handleShowCreateAccount(createAccountProps) {
+    if (this.siteSettings.enable_discourse_connect) {
+      const returnPath = encodeURIComponent(window.location.pathname);
+      window.location = getURL("/session/sso?return_path=" + returnPath);
+    } else {
+      if (this.isOnlyOneExternalLoginMethod) {
+        // we will automatically redirect to the external auth service
+        this.login.externalLogin(this.externalLoginMethods[0], {
+          signup: true,
+        });
+      } else {
+        this.router.transitionTo("signup").then((login) => {
+          Object.keys(createAccountProps || {}).forEach((key) => {
+            login.controller.set(key, createAccountProps[key]);
+          });
+        });
+      }
+    }
+  }
+
+  @action
+  showNotActivated(props) {
+    this.modal.show(NotActivatedModal, { model: props });
+  }
+
+  @action
+  async triggerLogin() {
+    if (this.loginDisabled) {
+      return;
+    }
+
+    if (isEmpty(this.loginName) || isEmpty(this.loginPassword)) {
+      this.flash = I18n.t("login.blank_username_or_password");
+      this.flashType = "error";
+      return;
+    }
+
+    try {
+      this.loggingIn = true;
+      const result = await ajax("/session", {
+        type: "POST",
+        data: {
+          login: this.loginName,
+          password: this.loginPassword,
+          second_factor_token:
+            this.securityKeyCredential || this.secondFactorToken,
+          second_factor_method: this.secondFactorMethod,
+          timezone: moment.tz.guess(),
+        },
+      });
```

Comment:
try/catch

Reply:
There's not a ton we can change here, but I prefer to keep those changes for another PR, just to ensure that if something breaks, it's not because of these changes